File: leetcode/C++/stacks/20.cpp
#include<bits/stdc++.h>
using namespace std;
//Cases: 
// ()[({}[])]
//   ^^
bool isValid(string s)
{
    stack<char> openB_stack, closeB_stack;
    for(int i = 0; i < s.length(); i++)
    {
        if(s[i] == '(' || s[i] == '{' || s[i] == '[')
            openB_stack.push(s[i]);
        else
        {
            //If there is a closed bracket before an open bracket
            if(openB_stack.empty())
                return false;
            //If is not the opposite bracket it will return false because the bracket is not in order
            if(!((openB_stack.top() ==  '(' && s[i] == ')' ) 
                 || (openB_stack.top() == '{' && s[i] == '}' )
                   || (openB_stack.top() == '[' && s[i] == ']' )))
                return false;
            //Remove the bracket from the stack
            openB_stack.pop();
        }
    }
    //If a bracket still in the stack it means there is an open bracket alone
    if(!openB_stack.empty())
        return false;
    return true;
}
int main(){
    string s;
    cin >> s;
    cout << "resultado: " << isValid(s) << endl;

    return 0;
}

File: leetcode/Go/stacks/20/20.go
package main

type Stack []byte

func (s *Stack) empty() bool {
	return len(*s) == 0
}
func (s *Stack) push(val byte) {
	*s = append(*s, val)
}

func (s *Stack) pop() byte {
	if s.empty() {
		return 'e'
	}
	top := (*s)[len(*s)-1]
	*s = (*s)[:len(*s)-1]
	return top
}

func isValid(s string) bool {
	var stk Stack
	if len(s) == 1 {
		return false
	}
	for i := range len(s) {
		// In case if there is an open bracket
		if s[i] == '(' || s[i] == '{' || s[i] == '[' {
			stk.push(s[i])
		} else {
			val := stk.pop()
			// Finding their pairs
			switch s[i] {
			case ')':
				if val != '(' {
					return false
				}
			case '}':
				if val != '{' {
					return false
				}
			case ']':
				if val != '[' {
					return false
				}
			default:
				// If there is no open bracket
				if val == 'e' {
					return false
				}
			}
		}
	}
	// The stack can be not empty in that case there is not a match with all open brackets
	if !stk.empty() {
		return false
	}
	return true
}
